SpatialGDK/Source/SpatialGDK/Private/Interop/SpatialReceiver.cpp

```diff
@@ -105,11 +105,44 @@ void USpatialReceiver::LeaveCriticalSection()
 		{
 			OnEntityAddedDelegate.Broadcast(PendingAddEntity);
 		}
+		PendingAddComponents.RemoveAll([PendingAddEntity](const PendingAddComponentWrapper& Component) {return Component.EntityId == PendingAddEntity;});
 	}
 
 	for (Worker_AuthorityChangeOp& PendingAuthorityChange : PendingAuthorityChanges)
 	{
-		HandleActorAuthority(PendingAuthorityChange);
+		if (PendingAuthorityChange.authority != WORKER_AUTHORITY_AUTHORITATIVE)
+		{
+			HandleActorAuthority(PendingAuthorityChange);
+		}
+	}
+
+	for (PendingAddComponentWrapper& PendingAddComponent : PendingAddComponents)
+	{
+		if (StaticComponentView->GetComponentData<UnrealMetadata>(PendingAddComponent.EntityId) == nullptr)
+		{
+			continue;
+		}
+		if (ClassInfoManager->IsGeneratedQBIMarkerComponent(PendingAddComponent.ComponentId))
+		{
+			continue;
+		}
+		if (StaticComponentView->HasAuthority(PendingAddComponent.EntityId, PendingAddComponent.ComponentId))
+		{
+			continue; // Hacky, allows servers to change state if they are going to be authoritative, without us overwriting it with old data
+		}
+
+		UE_LOG(LogSpatialReceiver, Verbose,
+			TEXT("Add component inside of a critical section, outside of an add entity, being handled: component id: %d, entity id: %d"),
+			PendingAddComponent.ComponentId, PendingAddComponent.EntityId);
+		HandleIndividualAddComponent(PendingAddComponent.EntityId, PendingAddComponent.ComponentId, MoveTemp(PendingAddComponent.Data));
+	}
+
+	for (Worker_AuthorityChangeOp& PendingAuthorityChange : PendingAuthorityChanges)
+	{
+		if (PendingAuthorityChange.authority == WORKER_AUTHORITY_AUTHORITATIVE)
```

Comment:
By the way, the reason the AuthorityChange processing is now split according to authority is to avoid cases where we receive data while being authoritative, as that could be unintuitive to the game devs. Now we process Lose Auth -> Add Components -> Gain Auth. A common thing that happens is that on handover we get ComponentData -> Gain Auth, and in the previous flow, we would have first gained authority and then applied updates. This is more intuitive, I think, you receive data as if you were a client to get the most up-to-date state, and then gain authority.

Reply:
This is a good comment - worth including it - or a subsection of it in the file

Reply:
This is only true for components in critical sections though, isn't it possible to get authority and a component data outside a critical section?

Reply:
You would then receive them in the correct order from the Worker SDK, and since we don't (shouldn't) queue them outside critical sections, it should be okay.